libs/platform/ebpf_object.c

```diff
@@ -22,59 +22,137 @@ static ebpf_lock_t _ebpf_object_tracking_list_lock = {0};
+    // Put the index in the high 2 bytes and the counter in the low 2 bytes.
+    // This allows us to more easily detect stale IDs while
+    // still keeping IDs in order by index.
+    return (index << 16) | _ebpf_id_table[index].counter;
+}
+
+// Get the index at which a given ID is stored.
+static inline ebpf_result_t
+_get_index_from_id(ebpf_id_t id, _Out_ uint32_t* index)
 {
+    uint32_t possible_index = id >> 16;
+    if (id != _get_id_from_index(possible_index)) {
+        return EBPF_KEY_NOT_FOUND;
+    }
+
+    *index = possible_index;
+    return EBPF_SUCCESS;
+}
+
+static ebpf_result_t
+_ebpf_object_tracking_list_insert(_Inout_ ebpf_object_t* object)
+{
+    int new_index;
+    ebpf_result_t return_value;
     ebpf_lock_state_t state;
     state = ebpf_lock_lock(&_ebpf_object_tracking_list_lock);
-    ebpf_list_insert_tail(&_ebpf_object_tracking_list, &object->global_list_entry);
+    for (new_index = 1; new_index < EBPF_COUNT_OF(_ebpf_id_table); new_index++) {
+        if (_ebpf_id_table[new_index].object == NULL) {
+            break;
+        }
+    }
+    if (new_index == EBPF_COUNT_OF(_ebpf_id_table)) {
+        return_value = EBPF_NO_MEMORY;
+        goto Done;
+    }
+
+    // Generate a new ID.
+    _ebpf_id_table[new_index].counter++;
+    _ebpf_id_table[new_index].object = object;
+    object->id = _get_id_from_index(new_index);
+
+    return_value = EBPF_SUCCESS;
+
+Done:
     ebpf_lock_unlock(&_ebpf_object_tracking_list_lock, state);
+
+    return EBPF_SUCCESS;
 }
 
 static void
 _ebpf_object_tracking_list_remove(ebpf_object_t* object)
 {
-    ebpf_lock_state_t state;
-    state = ebpf_lock_lock(&_ebpf_object_tracking_list_lock);
-    ebpf_list_remove_entry(&object->global_list_entry);
+    ebpf_lock_state_t state = ebpf_lock_lock(&_ebpf_object_tracking_list_lock);
+
+    uint32_t index;
+    ebpf_result_t return_value = _get_index_from_id(object->id, &index);
+    ebpf_assert(return_value == EBPF_SUCCESS);
+
+    // In a release build, ebpf_assert is a no-op so we
+    // need to avoid an unreferenced variable warning.
+    UNREFERENCED_PARAMETER(return_value);
+
+    _ebpf_id_table[index].object = NULL;
+
     ebpf_lock_unlock(&_ebpf_object_tracking_list_lock, state);
 }
 
 void
 ebpf_object_tracking_initiate()
 {
     ebpf_lock_create(&_ebpf_object_tracking_list_lock);
-    ebpf_list_initialize(&_ebpf_object_tracking_list);
+    memset(_ebpf_id_table, 0, sizeof(_ebpf_id_table));
 }
 
 void
 ebpf_object_tracking_terminate()
 {
-    ebpf_assert(ebpf_list_is_empty(&_ebpf_object_tracking_list));
+    for (int index = 0; index < EBPF_COUNT_OF(_ebpf_id_table); index++) {
+        ebpf_assert(_ebpf_id_table[index].object == NULL);
+    }
 }
 
-void
+ebpf_result_t
 ebpf_object_initialize(
     ebpf_object_t* object,
     ebpf_object_type_t object_type,
     ebpf_free_object_t free_function,
     ebpf_object_get_program_type_t get_program_type_function)
 {
+    ebpf_result_t result = _ebpf_object_tracking_list_insert(object);
```

Comment:
Here we are insertingthe ebpf_object in the global tracking list _before_ initializing the object, and we are not holding any lock while we initialize the object below. This can cause corruption if another thread is iterating over the global tracking list and tries to read/write any fields of ebpf_object.

I think inserting in the tracking list should be the last step.

Reply:
Done